posts/2021-11-30-microprofile-5-22001-beta.adoc

```diff
@@ -0,0 +1,203 @@
+
+
+[#jpa_persistence]
+=== Define JPA persistence properties at server scope
+
+This new JPA enhancement adds the ability to declare "default" JPA persistence properties, to all container-managed persistence contexts, as a Liberty server.xml configuration.
+
+Previously, if a persistence property needed to be set for all persistence.xml configuration files, you would manually update all persistence.xml files in all applications. This could end up requiring hundreds of manual updates and/or rebuilding of applications.
+With this enhancement, you can specify persistence properties, in the server.xml, that will propagate to all container-managed persistence units for applications installed on that server.
+
+In order to start using the new feature, you only need to add the new `<defaultProperties>` configuration to your `<jpa>` configuration, in server.xml, along with the persistence properties that you want to apply to all container-managed persistence units:
+
+===== Example 1:
+
+[source, xml]
+----
+<server>
+    <featureManager>
+        <feature>persistenceContainer-3.0</feature>
+    </featureManager>
+
+    <jpa defaultPersistenceProvider="org.hibernate.jpa.HibernatePersistenceProvider">
+        <defaultProperties>
+            <property name="javax.persistence.lock.timeout" value="4000"/>
+            <property name="hibernate.dialect" value="org.hibernate.dialect.Oracle12cDialect"/>
+        </defaultProperties>
+    </jpa>
+</server>
+----
+
+===== Example 2:
+
+[source, xml]
+----
+<server>
+    <featureManager>
+        <feature>jpa-2.1</feature>
+    </featureManager>
+
+    <jpa>
+        <defaultProperties>
+            <property name="javax.persistence.lock.timeout" value="12345"/>
+            <property name="eclipselink.cache.shared.default" value="false"/>
+        </defaultProperties>
+    </jpa>
+</server>
+----
+
+==== Technical description
+
+These `defaultProperties`, integration-level persistence properties will be supplied to the persistence provider specified when the `PersistenceProvider.createContainerEntityManagerFactory` method is called by the JPA Container.
+
+According to the JPA specification (3.4.4.3):
+
+```
+If the same property or hint is specified more than once, the following order of overriding applies, in order of decreasing precedence:
+  • argument to method of EntityManager, Query, or TypedQuery interface
+  • specification to NamedQuery (annotation or XML)
+  • argument to createEntityManagerFactory method
+  • specification in persistence.xml
+```
+
+This means these new `defaultProperties` persistence property values will override the properties specified in persistence.xml with the same name. However, property values specified through PersistenceContext annotation, or the `persistence-context-ref` deployment descriptor element, or Query Hints property will override these `defaultProperties`.
```

Comment:
These `defaultProperties` persistence property values override any  properties  with the same name that are specified in a `persistence.xml` file.